client/android/WorldScope/app/src/main/java/com/litmus/worldscope/LitmusPlayer.java

```diff
@@ -0,0 +1,284 @@
+import com.google.android.exoplayer.audio.AudioTrack;
+import com.google.android.exoplayer.drm.MediaDrmCallback;
+import com.google.android.exoplayer.drm.StreamingDrmSessionManager;
+import com.google.android.exoplayer.upstream.BandwidthMeter;
+import com.google.android.exoplayer.upstream.DefaultBandwidthMeter;
+import com.google.android.exoplayer.util.PlayerControl;
+
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class LitmusPlayer implements
+        ExoPlayer.Listener,
+        MediaCodecVideoTrackRenderer.EventListener,
+        MediaCodecAudioTrackRenderer.EventListener, DefaultBandwidthMeter.EventListener,
+        StreamingDrmSessionManager.EventListener {
+
+    public interface PlayerEventsListener {
+        void onVideoSizeChanged(int width, int height, int unappliedRotationDegrees,
+                                float pixelWidthHeightRatio);
+    }
+
+    private final String TAG = "LitmusPlayer";
+
+    private ExoPlayer player;
+    private DashRendererBuilder dashRendererBuilder;
+    private WidevineTestMediaDrmCallback drmCallback;
+    private String mpdLink;
+    private DashRendererBuilder rendererBuilder;
+    private PlayerControl playerControl;
+    private TrackRenderer videoRenderer;
+    private MediaCodecAudioTrackRenderer audioRenderer;
+    private Surface surface;
+    private CodecCounters codecCounters;
+    private CopyOnWriteArrayList<PlayerEventsListener> listeners;
+    private boolean surfaceLock;
+    private boolean rendererLock;
+
+    // Constants for instantiating ExoPlayer
+    private static final int MIN_BUFFER_MS = 1000;
+    private static final int MIN_REBUFFER_MS = 1000;
+
+    public static final int RENDERER_COUNT = 2;
+    public static final int SURFACE_READY = 0;
+    public static final int RENDERER_READY = 1;
+    public static final int TYPE_VIDEO = 0;
+    public static final int TYPE_AUDIO = 1;
+    public static final int TYPE_TEXT = 2;
+    public static final int TYPE_METADATA = 3;
+
+    // Constants pulled into this class for convenience.
+    public static final int STATE_IDLE = ExoPlayer.STATE_IDLE;
+    public static final int STATE_PREPARING = ExoPlayer.STATE_PREPARING;
+    public static final int STATE_BUFFERING = ExoPlayer.STATE_BUFFERING;
+    public static final int STATE_READY = ExoPlayer.STATE_READY;
+    public static final int STATE_ENDED = ExoPlayer.STATE_ENDED;
+    public static final int TRACK_DISABLED = ExoPlayer.TRACK_DISABLED;
+    public static final int TRACK_DEFAULT = ExoPlayer.TRACK_DEFAULT;
+
+    // Locks and conditions for push function to execute
+    final Lock pushLock = new ReentrantLock();
+    final Condition noSurface = pushLock.newCondition();
+    final Condition noRenderers = pushLock.newCondition();
+
+    // Handler given to DashRendererBuilder
+    private Handler mainHandler;
```

Comment:
You used surfaceLock and rendererLock instead  of Conditions and Locks used to solve the async issue?